File: app/src/main/java/com/shenhua/zhidaodaily/view/DailyAdapter.java
package com.shenhua.zhidaodaily.view;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.shenhua.zhidaodaily.R;
import com.shenhua.zhidaodaily.core.bean.BannerBean;
import com.shenhua.zhidaodaily.core.bean.HomeBean;
import com.shenhua.zhidaodaily.ui.DetailActivity;

import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Created by Shenhua on 12/13/2016.
 * e-mail shenhuanet@126.com
 *
 * @author shenhua
 */
public class DailyAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private LayoutInflater mLayoutInflater;
    private static final int ITEM_TYPE_HEADER = 0;
    private static final int ITEM_TYPE_CONTENT = 1;
    private Context mContext;
    private List<T> datas;

    public DailyAdapter(Context mContext, List<T> datas) {
        this.mContext = mContext;
        this.datas = datas;
        mLayoutInflater = LayoutInflater.from(mContext);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == ITEM_TYPE_HEADER) {
            return new HeaderViewHolder(mLayoutInflater.inflate(R.layout.view_logo, parent, false));
        } else if (viewType == ITEM_TYPE_CONTENT) {
            return new ContentViewHolder(mLayoutInflater.inflate(R.layout.list_item, parent, false));
        } else {
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof HeaderViewHolder) {
            final BannerBean bannerBean = (BannerBean) datas.get(0);
            final HeaderViewHolder headerViewHolder = (HeaderViewHolder) holder;
            headerViewHolder.titleTv.setText(bannerBean.getTitle());
            headerViewHolder.formTv.setText(bannerBean.getFrom());
            headerViewHolder.periods.setText(bannerBean.getPeriods());
            Glide.with(mContext).load(bannerBean.getImg()).placeholder(R.drawable.daily_bg2)
                    .error(R.drawable.noimage).centerCrop().into(headerViewHolder.logoIv);
            headerViewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Intent intent = new Intent(mContext, DetailActivity.class);
                    intent.putExtra("dataUrl", bannerBean.getHref());
                    intent.putExtra("title", bannerBean.getTitle());
                    intent.putExtra("img", bannerBean.getImg());
                    navToDetail(intent, headerViewHolder.logoIv);
                }
            });
        }

        if (holder instanceof ContentViewHolder) {
            List<HomeBean> homeBeanList = (List<HomeBean>) datas.get(1);
            final HomeBean bean = homeBeanList.get(position - 1);
            final ContentViewHolder contentViewHolder = (ContentViewHolder) holder;
            contentViewHolder.title.setText(bean.getTitle());
            contentViewHolder.detail.setText(bean.getDetail());
            contentViewHolder.read.setText(bean.getRead());
            Glide.with(mContext).load(bean.getImg()).placeholder(R.drawable.daily_bg2)
                    .error(R.drawable.noimage).centerCrop().into(contentViewHolder.image);
            contentViewHolder.layout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(mContext, DetailActivity.class);
                    intent.putExtra("dataUrl", bean.getHref());
                    intent.putExtra("title", bean.getTitle());
                    intent.putExtra("img", bean.getImg());
                    navToDetail(intent, contentViewHolder.image);
                }
            });
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return ITEM_TYPE_HEADER;
        } else {
            return ITEM_TYPE_CONTENT;
        }
    }

    public void setDatas(List<T> datas) {
        this.datas = datas;
        notifyDataSetChanged();
    }

    @SuppressWarnings("unchecked")
    public void addItem(HomeBean bean, int position) {
        datas.add(position, (T) bean);
        notifyItemInserted(position);
    }

    public void addMore() {

    }

    @SuppressWarnings("unchecked")
    @Override
    public int getItemCount() {
        if (datas == null || datas.size() == 0) {
            return 0;
        }
        return ((List<HomeBean>) datas.get(1)).size() + 1;
    }

    static class HeaderViewHolder extends RecyclerView.ViewHolder {

        @Bind(R.id.iv_logo)
        ImageView logoIv;
        @Bind(R.id.tv_periods)
        TextView periods;
        @Bind(R.id.tv_title)
        TextView titleTv;
        @Bind(R.id.tv_from)
        TextView formTv;

        HeaderViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    static class ContentViewHolder extends RecyclerView.ViewHolder {

        @Bind(R.id.layout)
        LinearLayout layout;
        @Bind(R.id.tv_title)
        TextView title;
        @Bind(R.id.tv_detail)
        TextView detail;
        @Bind(R.id.tv_read)
        TextView read;
        @Bind(R.id.iv_image)
        ImageView image;

        ContentViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    private void navToDetail(Intent intent, View imgView) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            ActivityOptions options = ActivityOptions
                    .makeSceneTransitionAnimation((Activity) mContext, imgView, "photos");
            mContext.startActivity(intent, options.toBundle());
        } else {
            ActivityOptionsCompat options = ActivityOptionsCompat
                    .makeScaleUpAnimation(imgView, imgView.getWidth() / 2, imgView.getHeight() / 2, 0, 0);
            ActivityCompat.startActivity(mContext, intent, options.toBundle());
        }
    }

}

File: app/src/main/java/com/shenhua/zhidaodaily/ui/DailyAdapter.kt
package com.shenhua.zhidaodaily.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.ViewCompat
import androidx.navigation.Navigation
import androidx.navigation.fragment.FragmentNavigator
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestOptions
import com.shenhua.zhidaodaily.R
import com.shenhua.zhidaodaily.core.Daily
import kotlinx.android.synthetic.main.item_daily_content.view.*
import kotlinx.android.synthetic.main.item_daily_header.view.*


/**
 * Created by shenhua on 2018/10/19.
 * @author shenhua
 *         Email shenhuanet@126.com
 */
class DailyAdapter : ListAdapter<Daily, RecyclerView.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == ITEM_TYPE_HEADER) {
            HeaderViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_daily_header, parent, false))
        } else {
            ContentViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.item_daily_content, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is HeaderViewHolder) {
            holder.bind(getItem(position))
        } else if (holder is ContentViewHolder) {
            holder.bind(getItem(position))
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) ITEM_TYPE_HEADER else super.getItemViewType(position)
    }

    inner class ContentViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(daily: Daily) {
            itemView.isClickable = true
            itemView.setOnClickListener {
                val extras = FragmentNavigator.Extras.Builder()
                        .addSharedElement(itemView.image, "photos").build()
                val bundle = Bundle()
                bundle.putSerializable("daily", daily)
                Navigation.findNavController(itemView).navigate(R.id.action_dailyFragment_to_contentFragment,
                        bundle,
                        null,
                        extras)
            }
            itemView.title.text = daily.title
            itemView.description.text = daily.description
            itemView.read.text = daily.read
            Glide.with(itemView.context).load(daily.image).apply(RequestOptions().placeholder(R.drawable.bg_img_mask).error(R.drawable.daily_bg))
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(itemView.image)
            itemView.tag = daily.link
        }
    }

    inner class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(daily: Daily) {
            itemView.isClickable = true
            itemView.setOnClickListener {
                val extras = FragmentNavigator.Extras.Builder()
                        .addSharedElement(itemView.iv_image, ViewCompat.getTransitionName(itemView.iv_image)!!).build()
                val bundle = Bundle()
                bundle.putSerializable("daily", daily)
                Navigation.findNavController(itemView).navigate(R.id.action_dailyFragment_to_contentFragment,
                        bundle,
                        null,
                        extras)
            }
            itemView.tv_title.text = daily.title
            itemView.tv_from.text = daily.from
            itemView.periods.text = String.format(itemView.resources.getString(R.string.string_periods_format), daily.periods, daily.data)
            Glide.with(itemView.context).load(daily.image).apply(RequestOptions().placeholder(R.drawable.daily_bg).error(R.drawable.daily_bg))
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(itemView.iv_image)
            itemView.tag = daily.link
        }
    }

    companion object {
        private const val ITEM_TYPE_HEADER = 0x12

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Daily>() {
            override fun areItemsTheSame(oldItem: Daily, newItem: Daily): Boolean {
                return oldItem == newItem
            }

            override fun areContentsTheSame(oldItem: Daily, newItem: Daily): Boolean {
                return oldItem == newItem
            }
        }
    }
}